materialize-boilerplate/stream-encode/parquet.go

```diff
@@ -0,0 +1,701 @@
+	return func(cfg *parquetConfig) {
+		cfg.rowGroupByteLimit = n
+	}
+}
+
+func NewParquetEncoder(w io.WriteCloser, sch ParquetSchema, opts ...ParquetOption) *ParquetEncoder {
+	cfg := parquetConfig{
+		compression:       Uncompressed,
+		rowGroupRowLimit:  defaultRowGroupRowLimit,
+		rowGroupByteLimit: defaultRowGroupByteLimit,
+	}
+	for _, o := range opts {
+		o(&cfg)
+	}
+
+	fields := make(schema.FieldList, 0, len(sch))
+	for _, e := range sch {
+		fields = append(fields, makeNode(e))
+	}
+
+	schemaRoot := schema.MustGroup(schema.NewGroupNode("schema", parquet.Repetitions.Required, fields, -1))
+
+	cwc := &countingWriteCloser{w: w}
+
+	return &ParquetEncoder{
+		cfg:        cfg,
+		schema:     sch,
+		schemaRoot: schemaRoot,
+		sinkWriter: file.NewParquetWriter(cwc, schemaRoot, writerOpts(cfg)),
+		cwc:        cwc,
+		rs:         newRowSizing(sch),
+	}
+}
+
+func writerOpts(cfg parquetConfig) file.WriteOption {
+	propOpts := []parquet.WriterProperty{}
+
+	switch cfg.compression {
+	case Uncompressed:
+		propOpts = append(propOpts, parquet.WithCompression(compress.Codecs.Uncompressed))
+	case Snappy:
+		propOpts = append(propOpts, parquet.WithCompression(compress.Codecs.Snappy))
+	case Gzip:
+		propOpts = append(propOpts, parquet.WithCompression(compress.Codecs.Gzip))
+	default:
+		panic(fmt.Sprintf("unknown compression setting: %d", cfg.compression))
+	}
+
+	return file.WithWriterProps(parquet.NewWriterProperties(propOpts...))
+}
+
+// Encode encodes a row of data by buffering it in the encoder'ss buffer, and if thresholds are
+// exceed writing the buffer as a row group to the scratch file, and potentially flushing the row
+// groups from the scratch file collectively as a single row group to the output.
+func (e *ParquetEncoder) Encode(row []any) error {
+	var err error
+
+	if e.scratch.writer == nil {
+		// Either the very first row, or the first one after flushing the scratch file.
+		if e.scratch.file, err = os.CreateTemp("", "parquet-scratch-*"); err != nil {
+			return fmt.Errorf("encode creating scratch file: %W", err)
+		}
+		e.scratch.writer = file.NewParquetWriter(e.scratch.file, e.schemaRoot)
+	}
+
+	e.buffer = append(e.buffer, row)
+	e.bufferSizeBytes += e.rs.estSize(row)
+	e.scratch.rowCount += 1
+
+	if e.bufferSizeBytes >= maxBufferSize {
+		// Write out the buffer as a single row group to the scratch file.
+		if err := e.flushBuffer(); err != nil {
+			return fmt.Errorf("encode flushing buffer based on buffer size: %W", err)
+		}
+	}
+
+	if e.scratch.sizeBytes >= e.cfg.rowGroupByteLimit ||
+		e.scratch.rowCount >= e.cfg.rowGroupRowLimit ||
+		e.scratch.columnChunkCount >= maxScratchColumnChunkCount {
+		if e.scratch.columnChunkCount >= maxScratchColumnChunkCount {
+			log.WithFields(log.Fields{
+				"scratchSizeBytes":        e.scratch.sizeBytes,
+				"scratchRowCount":         e.scratch.rowCount,
+				"scratchColumnChunkCount": e.scratch.columnChunkCount,
+			}).Debug("flushing scratch file based on column chunk count")
```

Comment:
for my learning: why do we specificaly log this one?

Reply:
It would mean that we'd be producing row groups that were smaller than the configured row group byte limit or row count limit. So it could be a troubleshooting mechanism in case somebody wonders why their row groups are too small.